### PR TITLE
[IMP] l10n_dk_oioubl: allow import

### PR DESCRIPTION
**Pull request by Levizar**
The aim of this commit is to allow import of OIOUBL xml files.

Additional notes:
some modification to the way ubl 2.0 retrieve tax node was required as it was unable to retrieve taxes on OIOUBL.

task-id: 3420748

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -3102,7 +3102,7 @@ def _extend_with_attachments(self, attachments, new=False):
         """
         success = False
 
-        for file_data in attachments._unwrap_edi_attachments(): # sorted by priority
+        for file_data in attachments._unwrap_edi_attachments():  # sorted by priority
```
Comment (smetl) [inline addons/account/models/account_move.py:3105]: useless diff

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py [hunk 2]
```diff
@@ -660,15 +660,16 @@ def _import_fill_invoice_line_form(self, tree, invoice_line, qty_factor):
         tax_nodes = tree.findall('.//{*}Item/{*}ClassifiedTaxCategory/{*}Percent')
         if not tax_nodes:
             for elem in tree.findall('.//{*}TaxTotal'):
-                tax_nodes += elem.findall('.//{*}TaxSubtotal/{*}Percent')
+                tax_node = elem.findall('.//{*}TaxSubtotal//{*}Percent')
+                tax_nodes += tax_node
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:664]: useless diff

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py [hunk 3]
```diff
@@ -660,15 +660,16 @@ def _import_fill_invoice_line_form(self, tree, invoice_line, qty_factor):
...
         return self._import_fill_invoice_line_taxes(tax_nodes, invoice_line, inv_line_vals, logs)
 
     def _correct_invoice_tax_amount(self, tree, invoice):
         """ The tax total may have been modified for rounding purpose, if so we should use the imported tax and not
          the computed one """
         # For each tax in our tax total, get the amount as well as the total in the xml.
         for elem in tree.findall('.//{*}TaxTotal/{*}TaxSubtotal'):
-            percentage = elem.find('.//{*}TaxCategory/{*}Percent')
+            percentage = elem.find('.//{*}TaxCategory//{*}Percent')
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:672]: useless diff

File: addons/l10n_account_edi_ubl_cii_tests/tests/common.py [hunk 4]
```diff
@@ -234,9 +234,9 @@ def _assert_invoice_attachment(self, attachment, xpaths, expected_file_path):
         xml_content = base64.b64decode(attachment.with_context(bin_size=False).datas)
         xml_etree = self.get_xml_tree_from_string(xml_content)
 
-        expected_file = get_resource_path(self.test_module, 'tests/test_files', expected_file_path)
-        self.assertTrue(expected_file, "expected file not found")
-        expected_etree = etree.parse(expected_file).getroot()
+        expected_file_full_path = get_resource_path(self.test_module, 'tests/test_files', expected_file_path)
+        self.assertTrue(expected_file_full_path, "expected file not found")
+        expected_etree = etree.parse(expected_file_full_path).getroot()
```
Comment (smetl) [inline addons/l10n_account_edi_ubl_cii_tests/tests/common.py:239]: useless diff

File: addons/l10n_dk_oioubl/models/account_move.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    @api.model
+    def _get_ubl_cii_builder_from_xml_tree(self, tree):
+        customization_id = tree.find('{*}CustomizationID')
+        if customization_id is not None and 'OIOUBL-2' in customization_id.text:
+            return self.env['account.edi.xml.oioubl_201']
+        super()._get_ubl_cii_builder_from_xml_tree(tree)
```
Comment (smetl) [inline addons/l10n_dk_oioubl/models/account_move.py:12]: You are breaking the super call! :angry: 
```suggestion
    def _get_ubl_cii_builder_from_xml_tree(self, tree):
        builder = super()._get_ubl_cii_builder_from_xml_tree(tree)
        if builder:
            return builder
        customization_id = tree.find('{*}CustomizationID')
        if customization_id is not None and 'OIOUBL-2' in customization_id.text:
            return self.env['account.edi.xml.oioubl_201']
```

File: addons/l10n_dk_oioubl/tests/test_files/external/ADVORD_01_01_00_Invoice_v2p1.xml [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+		For more information, see www.oioubl.dk	or email oioubl@itst.dk
+		
+******************************************************************************************************************
+-->
+<!-- The following TestInstance process instruction is for customization usage. It indicates that the instance is for testing purposes -->
+<?TestInstance ResponseTo="smtp:test@company.dk"
+	description= "apply your comment here"
+?>
+
+	<cbc:UBLVersionID>2.0</cbc:UBLVersionID>
```
Comment (smetl) [inline addons/l10n_dk_oioubl/tests/test_files/external/ADVORD_01_01_00_Invoice_v2p1.xml:31]: indent is very louche :detective:

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py (hunk 2 above)
Comment (Levizar) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:664]: Make it back to what it was and try to make the test pass without this. 😁

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py (hunk 3 above)
Comment (Levizar) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:672]: No, it's not: Try to make my tests pass without it.

File: addons/l10n_dk_oioubl/tests/test_files/external/ADVORD_01_01_00_Invoice_v2p1.xml (hunk 6 above)
Comment (Levizar) [reply addons/l10n_dk_oioubl/tests/test_files/external/ADVORD_01_01_00_Invoice_v2p1.xml:31]: It's the official xmls, I didn't make any modification to them.

File: addons/l10n_dk_oioubl/models/account_move.py (hunk 5 above)
Comment (Levizar) [reply addons/l10n_dk_oioubl/models/account_move.py:12]: Why is it an issue?
It only breaks it if and only if it's an OIOUBL format.

Comment (smetl) [reply addons/l10n_dk_oioubl/models/account_move.py:12]: @Levizar you missed to return the super results

File: addons/l10n_dk_oioubl/tests/test_files/external/ADVORD_01_01_00_Invoice_v2p1.xml (hunk 6 above)
Comment (smetl) [reply addons/l10n_dk_oioubl/tests/test_files/external/ADVORD_01_01_00_Invoice_v2p1.xml:31]: @Levizar indent it using 4 whitespaces at least :(

File: addons/l10n_dk_oioubl/models/account_move.py (hunk 5 above)
Comment (smetl) [reply addons/l10n_dk_oioubl/models/account_move.py:12]: @Levizar You need at least to do:
```suggestion
    def _get_ubl_cii_builder_from_xml_tree(self, tree):
        customization_id = tree.find('{*}CustomizationID')
        if customization_id is not None and 'OIOUBL-2' in customization_id.text:
            return self.env['account.edi.xml.oioubl_201']
        return super()._get_ubl_cii_builder_from_xml_tree(tree)
```

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py (hunk 3 above)
Comment (smetl) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:672]: @Levizar Done... no error :laughing:

Comment (Levizar) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:672]: If you undo this one, you also have to undo the previous one:
image

And when you do it, you get this:
image

File: addons/l10n_account_edi_ubl_cii_tests/tests/common.py (hunk 4 above)
Comment (Levizar) [reply addons/l10n_account_edi_ubl_cii_tests/tests/common.py:239]: I disagree: I put a wrong and misleading variable name and it needs to be corrected, except if you consider it's not an issue to have wrong misleading variable name.

File: addons/l10n_dk_oioubl/models/account_move.py (hunk 5 above)
Comment (Levizar) [reply addons/l10n_dk_oioubl/models/account_move.py:12]: You are right for that! I haven't seen it. Thanks!

But I'm a bit worried: this bug didn't break any test. :grimacing:

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py (hunk 3 above)
Comment (Levizar) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:747]: My bad, I understand what you mean.
You meant only the `+=` is useless diff previously but on tax category we always get the percent

File: addons/l10n_account_edi_ubl_cii_tests/tests/common.py (hunk 4 above)
Comment (Levizar) [reply addons/l10n_account_edi_ubl_cii_tests/tests/common.py:239]: IRL: we decided to put it in another commit to not pollute the original commit with the diff

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py (hunk 2 above)
Comment (Levizar) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py]: We decided to put that part in another commit as it's a fix by itself for UBL 2.0

File: addons/l10n_dk_oioubl/tests/test_files/external/ADVORD_01_01_00_Invoice_v2p1.xml (hunk 6 above)
Comment (Levizar) [reply addons/l10n_dk_oioubl/tests/test_files/external/ADVORD_01_01_00_Invoice_v2p1.xml:31]: IRL: We decided to keep the xml as-is provided by the official documentation

**Pull request closed without merging**